concept/thing/thing_manager.rs

```diff
@@ -2943,23 +2929,47 @@ impl ThingManager {
             .map_ok(|(roleplayer, _count)| (roleplayer.player(), roleplayer.role_type()));
         for rp in players {
             let (rp_player, rp_role_type) = rp?;
-            debug_assert!(!(rp_player == Object::new(player.vertex()) && role_type == rp_role_type));
+
             let index = ThingEdgeIndexedRelation::new(
                 player.vertex(),
                 rp_player.vertex(),
                 relation.vertex(),
                 role_type.vertex().type_id_(),
                 rp_role_type.vertex().type_id_(),
-            );
-            snapshot.delete(index.into_storage_key().into_owned_array());
+            )
+            .into_storage_key();
+            let snapshot_index_value_opt = Self::get_snapshot_put_value(snapshot, index.as_reference());
+            let index_array = index.clone().into_owned_array();
+            if let Some(snapshot_index_value) = snapshot_index_value_opt {
+                snapshot.unput_val(index_array.clone(), snapshot_index_value);
+            }
+
             let index_reverse = ThingEdgeIndexedRelation::new(
                 rp_player.vertex(),
                 player.vertex(),
                 relation.vertex(),
                 rp_role_type.vertex().type_id_(),
                 role_type.vertex().type_id_(),
-            );
-            snapshot.delete(index_reverse.into_storage_key().into_owned_array());
+            )
+            .into_storage_key();
+            let snapshot_index_reverse_value_opt = Self::get_snapshot_put_value(snapshot, index_reverse.as_reference());
+            let index_reverse_array = index_reverse.into_owned_array();
+            if let Some(snapshot_index_reverse_value) = snapshot_index_reverse_value_opt {
+                snapshot.unput_val(index_reverse_array.clone(), snapshot_index_reverse_value);
+            }
+
+            let is_persisted = snapshot
+                .get_mapped(index.as_reference(), |_| true)
+                .map_err(|err| Box::new(ConceptReadError::SnapshotGet { source: err }))?
+                .unwrap_or(false);
```

Comment:
Small nit, I think this reads better with `.is_some()`
```suggestion
            let is_persisted = snapshot
                .get_mapped(index.as_reference(), |_| ())
                .map_err(|err| Box::new(ConceptReadError::SnapshotGet { source: err }))?
                .is_some();
```

Reply:
True, it was a little copypaste